LandSandBoat/server
# [LUA] Provide more robust casting behavior for light spirit

**_I affirm:_**

## What does this pull request do?

I've not found a more thorough outline of summoner spirit behavior than [this guide](https://ffxiclopedia.fandom.com/wiki/Spirit_Guide_by_Spira#Elemental_Spirit_Casting_Timer), so this is what i'm using to base the code on

This PR piggybacks on https://github.com/LandSandBoat/server/pull/4891 (which may need some additional tweaks to cleanly support this code) and does the following:
- creates a listener on the spirit pact's owner for tick and job ability
  - job ability `assault` and `retreat` will reset the casting cooldown and undo the "half cooldown" as well
  - tick listener will refresh the pet's casting cooldown based on the stipulations outlined in the `setMagicCastCooldown` function
- returns `-1` in `onMobMagicPrepare` when the logic determines the spirit shouldn't cast at all from `mob_controller`, which casts as `uint` and forces an early exit out of `TryCastSpell`. This is what may need to be updated in the other PR, as it results in debug warnings in the map server log. However, this behavior is done so that we can more granularly choose spell and target for the spirit pacts. The behavior is used to
  - stop the spirit from casting immediately upon spawn 
  - cast heals/buffs on particular party/alliance members
  - then return `-1` to make `TryCastSpell` not use generic `mob_spell_list` logic to cast a spell
- does some complex logic for light spirit:
  - loops over the whole alliance and finds the people who are yellow/orange/red hp, but doesn't guarantee healing them (tiered chance based on how low the hp is)
  - if a player is chosen to heal, random chance to use either highest cure or curaga
  - if no `cureTarget` is chosen, loops over the player's party to attempt buffs
  - In Buffing Mode, Light Spirit's spirit timer is halved.
    - This is done via a localvar. Utilized when pet is first summoned and after casting an enhancing magic spell
  - There are only 4 types of spells Light Spirit can cast (i.e Protect, Shell, Regen and Haste). Light Spirit cannot cast -ra buffs, bar-spells or -na spells.
  - Light Spirit will overwrite lower tiered Protect and Shell spells but will not overwrite Haste or Regen.
  - Light Spirit does not cast buffs on players who are under the Invisible status.
  - Light Spirit cannot cast buffs on itself.
  - Light Spirit can only cast Regen I and not II or III
  - Light Spirit can be encouraged to cast its buff spells on party members by having it face towards a party member.
    - This is calculated at the same time as the logic for finding a `buffTarget`
    - `buffTarget` is set to `nil` if the light spirit is in combat and the delicate positioning isn't setup

## Steps to test these changes

All spirit pacts are affected by the spell casting delay changes
- summon thunder spirit, aggro mob and see that it very quickly will cast a spell
- resummon thunder spirit and `assault` to see that it takes longer (twice as long) to cast a spell
- afaik none of the non-lightspirit pacts have buff spells of any kind, but if they have them defined in their `mob_spell_list` then they'll use them while idle

Light spirit testing is a bit more involved
- build a party/alliance
- set hp low on a few members. see that is will, on average, heal the lower HP player (but not always, and sometimes won't heal anyone even with multiple in yellow hp)
- position a party member between master and light spirit to see that it will always choose that player to buff (if it doesn't cure anyone)
- set the light spirit to engage a mob, see that it mostly casts nukes (same healing behavior as above though). It will also buff if a player is in the very particular position between light spirit and mob
  - > In addition to the spell repertoire for its offensive mode, the Light Spirit can still cast buffing mode spells while it is engaged.
